rust-query-crlite/src/main.rs

```diff
@@ -0,0 +1,794 @@
+        Ok(rustls::client::ServerCertVerified::assertion())
+    }
+}
+
+fn signoff(db: &CRLiteDB, host_file_url: &str) -> Result<CmdResult, CRLiteDBError> {
+    let host_file = reqwest::blocking::get(host_file_url)
+        .map_err(|_| CRLiteDBError::from("could not fetch hosts file"))?
+        .text()
+        .map_err(|_| CRLiteDBError::from("could not read hosts file"))?;
+
+    let hosts: Vec<&str> = host_file
+        .lines()
+        .map(|line| line.trim())
+        .filter(|line| !(line.starts_with('#') || line.is_empty()))
+        .collect();
+
+    query_https_hosts(db, &hosts)
+}
+
+fn query_https_hosts(db: &CRLiteDB, hosts: &[&str]) -> Result<CmdResult, CRLiteDBError> {
+    let mut found_revoked_certs = false;
+
+    let config = Arc::new(
+        rustls::ClientConfig::builder()
+            .with_safe_defaults()
+            .with_custom_certificate_verifier(Arc::new(AcceptAllCertsVerifier))
+            .with_no_client_auth(),
+    );
+
+    'host_loop: for host in hosts.iter() {
+        let (host, port) = host.rsplit_once(':').unwrap_or((host, "443"));
+        let port: u16 = port
+            .parse()
+            .map_err(|_| CRLiteDBError::from(format!("{}: malformed host port", host)))?;
+        let addrs = (String::from(host), port)
+            .to_socket_addrs()
+            .map_err(|e| CRLiteDBError::from(format!("could not lookup {}: {}", host, e)))?;
+        for addr in addrs {
+            let server_name = rustls::ServerName::try_from(host)
+                .map_err(|_| CRLiteDBError::from(format!("invalid DNS name: {}", host)))?;
+
+            let mut conn = rustls::ClientConnection::new(Arc::clone(&config), server_name)
+                .map_err(|e| CRLiteDBError::from(format!("{}: tls error: {}", host, e)))?;
+
+            let mut sock = TcpStream::connect(addr)
+                .map_err(|e| CRLiteDBError::from(format!("{}: tcp error: {}", host, e)))?;
+
+            let mut tls = rustls::Stream::new(&mut conn, &mut sock);
+            let _ = tls.write_all("HEAD / HTTP/1.1\r\n\r\n".as_bytes());
+
+            if let Some(certs) = conn.peer_certificates() {
+                if let Ok((_, cert)) = X509Certificate::from_der(certs[0].as_ref()) {
+                    debug!("Loaded certificate from {}", host);
+                    let status = db.query(&cert);
+                    match status {
+                        Status::Expired => warn!("{} {:?}", host, status),
+                        Status::Good => info!("{} {:?}", host, status),
+                        Status::NotCovered => warn!("{} {:?}", host, status),
+                        Status::NotEnrolled => warn!("{} {:?}", host, status),
+                        Status::Revoked => {
+                            found_revoked_certs = true;
+                            error!("{} {:?}", host, status);
+                        }
+                    }
+                    continue 'host_loop; // no need to check further addrs
```

Comment:
Rather than have a big outer loop, maybe there could be a function to check just one host at a time.